### PR TITLE
Scroll improvements:

### PR DESCRIPTION
**Pull request by rtrg**
Uncontrolled scrollbars, using [react-custom-scrollbars](https://github.com/malte-wessel/react-custom-scrollbars).
Avoid unnecessary rendering when scrolling stops.

## Motivation and Context
Updating scrollbar position in controlled way is causing scroll to lag (as it has to wait for its siblings to be updated), using [react-custom-scrollbars](https://github.com/malte-wessel/react-custom-scrollbars) (which maintains it scrolling state internally and fires required events) we can make scroll bar to be uncontrolled.

## How Has This Been Tested?
Tested manually

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: src/FixedDataTable.js [hunk 1]
```diff
@@ -437,7 +437,7 @@ class FixedDataTable extends React.Component {
   }
 
   componentWillMount() {
-    this._didScrollStop = debounceCore(this._didScrollStopSync, 200, this);
+    this._didScrollStop = debounceCore(this._didScrollStopSync, 50, this);
```
Comment (pradeepnschrodinger) [inline src/FixedDataTable.js]: Should we move the delay to a configurable prop, or for now to a constant

File: src/FixedDataTable.js [hunk 2]
```diff
@@ -567,6 +567,7 @@ class FixedDataTable extends React.Component {
 
   componentDidMount() {
     this._reportContentHeight();
+    this._didControlledScroll(this.props);
```
Comment (pradeepnschrodinger) [inline src/FixedDataTable.js]: why do we need this?

File: src/FixedDataTable.js [hunk 3]
```diff
@@ -818,11 +840,11 @@ class FixedDataTable extends React.Component {
 
   _renderRows = (/*number*/ offsetTop, fixedCellTemplates, fixedRightCellTemplates, scrollableCellTemplates,
     bodyHeight) /*object*/ => {
-    const { scrollEnabledY, scrolling } = scrollbarsVisible(this.props);
+    const { scrollEnabledY } = scrollbarsVisible(this.props);
     const props = this.props;
     return (
       <FixedDataTableBufferedRows
-        isScrolling={scrolling}
+        isScrolling={props.scrolling}
```
Comment (pradeepnschrodinger) [inline src/FixedDataTable.js]: thanks for fixing this! (part discussion in #399)

File: src/FixedDataTable.js [hunk 4]
```diff
@@ -932,25 +953,23 @@ class FixedDataTable extends React.Component {
...
-      onVerticalScroll,
-      scrollActions,
       scrollFlags,
       scrollX,
       scrollY,
       scrolling,
     } = this.props;
     const { overflowX, overflowY } = scrollFlags;
 
+    this._didScrollStop.reset();
```
Comment (pradeepnschrodinger) [inline src/FixedDataTable.js]: Is there any reason to reset this early on? Also wouldn't the this_.scrollToX/this._scrollToY calls below call this._onHorizontalScroll/this._onVerticalScroll which would in turn reset it again?

File: src/FixedDataTable.js [hunk 5]
```diff
@@ -932,25 +953,23 @@ class FixedDataTable extends React.Component {
...
     if (!scrolling) {
       this._didScrollStart();
```
Comment (pradeepnschrodinger) [inline src/FixedDataTable.js]: Also I'm thinking if we should check if an actual scroll resulted from this and then call this._didScrollStart and this._didScrollEnd (the if's down below may not be true always)

File: src/FixedDataTable.js [hunk 6]
```diff
@@ -1096,12 +1132,12 @@ class FixedDataTable extends React.Component {
 
     if (scrollJumpedX) {
       scrollActions.jumpScrollX();
-      onHorizontalScroll && onHorizontalScroll(scrollX);
+      this._scrollToX(scrollX);
```
Comment (pradeepnschrodinger) [inline src/FixedDataTable.js]: A jump scroll is detected happens `after` the actual scroll happened (because it's instantaneous). Wouldn't this be just making a scroll call again? (Same down below)

File: src/FixedDataTableCellGroup.js [hunk 7]
```diff
@@ -192,14 +192,6 @@ class FixedDataTableCellGroup extends React.Component {
     zIndex: PropTypes.number.isRequired,
   }
 
-  shouldComponentUpdate(/*object*/ nextProps) /*boolean*/ {
```
Comment (pradeepnschrodinger) [inline src/FixedDataTableCellGroup.js]: We still need this check and added parent's check (FixedDataTableRowImpl) doesn't cover it. 
This is because there can be multiple cell groups rendered by a row, and the scroll value can be different/same for each group (fixed cellgroups)

File: src/FixedDataTableCell.js [hunk 8]
```diff
@@ -91,7 +91,7 @@ class FixedDataTableCell extends React.Component {
   }
 
   shouldComponentUpdate(nextProps) {
-    if (nextProps.isScrolling && this.props.rowIndex === nextProps.rowIndex) {
+    if ((nextProps.isScrolling || this.props.isScrolling) && this.props.rowIndex === nextProps.rowIndex) {
```
Comment (pradeepnschrodinger) [inline src/FixedDataTableCell.js]: Wouldn't we miss the update with `props.isScrolling` turned on to `false`, and the cell will never know about it.
I think some folks require to know about this flag (see recent [PR comment](https://github.com/schrodinger/fixed-data-table-2/pull/398#discussion_r261081942)).

Also is it really worth skipping a single update for this?

File: src/FixedDataTableRow.js [hunk 9]
```diff
@@ -153,6 +152,17 @@ class FixedDataTableRowImpl extends React.Component {
     touchEnabled: PropTypes.bool,
   };
 
+  shouldComponentUpdate(/*object*/ nextProps) /*boolean*/ {
+    // when scrolling stops, nextProps.isScrolling will be false, so we use this.props.isScrolling to avoid unnecessary rendering
+    // where row index is same during vertical scrolling
+    return (
+        (!nextProps.isScrolling && !this.props.isScrolling) ||
```
Comment (pradeepnschrodinger) [inline src/FixedDataTableRow.js]: Same as the above comment about the scrolling part.
Also can we use 
```js
!(nextProps.isScrolling || this.props.isScrolling)
```

Btw, will this skip other updates to the row props, such as width, rowHeight, columnReorderingData, etc. while scrolling? Not sure if they could lead to problems.

File: src/FixedDataTable.js (hunk 1 above)
Comment (rtrg) [reply src/FixedDataTable.js]: moved to constant

File: src/FixedDataTable.js (hunk 2 above)
Comment (rtrg) [reply src/FixedDataTable.js]: I assumed that didControlledScroll actually is supposed to trigger scroll, when scrollX and scrollY is present during mount I want it to be propagated to scrollbar as it is uncontrolled now.

Made changes according, now that I know `_didControlledScroll` is just triggering callbacks

File: src/FixedDataTable.js (hunk 4 above)
Comment (rtrg) [reply src/FixedDataTable.js]: because this function will always trigger stop again, so resetting existing one to avoid unnecessary renders bcs of  `isScrolling false flag` when scroll is done in quick successions.

File: src/FixedDataTable.js (hunk 6 above)
Comment (rtrg) [reply src/FixedDataTable.js]: `scrollActions.jumpScrollX()` made me think it is triggering scroll, reverted these changes now that I know we are just triggering callbacks in this function

File: src/FixedDataTableCellGroup.js (hunk 7 above)
Comment (rtrg) [reply src/FixedDataTableCellGroup.js]: yes agreed! Made changes

File: src/FixedDataTableCell.js (hunk 8 above)
Comment (rtrg) [reply src/FixedDataTableCell.js]: Updates are missed even during scrolling, I feel we should group what prop we care about and then check them too along with isScrolling check.

Without this change every cell present in view port will be rendered once per scroll stop. If I have large number of rows and columns (100 rows and 10 columns ), and I quickly scroll multiple times (3 times) effectively 100 * 10 * 3 cells will be rendered.

Once migration to react 16 is done, can use context to provide scroll info to cells.

**Comment by rtrg** [comment]
I made necessary changes and also updated failing test cases bcs of scroll size changes.

File: src/FixedDataTable.js (hunk 6 above)
Comment (pradeepnschrodinger) [reply src/FixedDataTable.js]: Apologies, we need the call to `this._scrollToX`. Your initial change was correct.

Earlier we didn't need a scroll call since scrolling was controlled. 
Here since it's uncontrolled, whenever a scroll happens outside of the scroll bar (which is exactly what jump scrolls are), we need to explicitly set the scroll.

Also, this shouldn't lead to extra renders since `this._onVerticalScroll` and `this._onHorizontalScroll` checks for value change before issuing scroll actions (which was my initial concern).

File: src/FixedDataTable.js [hunk 10]
```diff
@@ -18,20 +18,21 @@ import FixedDataTableColumnResizeHandle from 'FixedDataTableColumnResizeHandle';
...
 import debounceCore from 'debounceCore';
 import isNaN from 'lodash/isNaN';
 import joinClasses from 'joinClasses';
 import scrollbarsVisible from 'scrollbarsVisible';
 import tableHeightsSelector from 'tableHeights';
+import {SCROLLBAR_STYLE} from 'scrollbar';
 
 var ARROW_SCROLL_SPEED = 25;
-
+const SCROLL_STOP_DEBOUNCE_TIME = 50;
```
Comment (pradeepnschrodinger) [inline src/FixedDataTable.js]: Hey, why was this changed from 50 to 200?

File: src/FixedDataTable.js (hunk 4 above)
Comment (pradeepnschrodinger) [reply src/FixedDataTable.js]: > this function will always trigger stop again
Yes, then doesn't it mean that the stop call is debounced again, thus avoiding the need of the reset call?

File: src/FixedDataTable.js [hunk 11]
```diff
@@ -962,10 +987,8 @@ class FixedDataTable extends React.Component {
       y = y < 0 ? 0 : y;
       y = y > maxScrollY ? maxScrollY : y;
 
-      //NOTE (jordan) This is a hacky workaround to prevent FDT from setting its internal state
-      if (onVerticalScroll ? onVerticalScroll(y) : true) {
-        scrollActions.scrollToY(y);
-      }
+
```
Comment (pradeepnschrodinger) [inline src/FixedDataTable.js]: nit: extra line

File: src/FixedDataTableCell.js (hunk 8 above)
Comment (pradeepnschrodinger) [reply src/FixedDataTableCell.js]: > Updates are missed even during scrolling

Yea, that's the current optimization. With your change updates are missed even right after scrolling.

Let's say that scrolling is stopped, then on the next render shouldComponentUpdate is called, but this.props.isScrolling is still true. This will mean that the cells don't get updated. This will save a render but it's a required render as the cells need to know when scrolling has stopped.

If the cells don't know when scrolling has stopped, it will lead to problems. 
It breaks the API (see [comment](https://github.com/schrodinger/fixed-data-table-2/pull/398#discussion_r261081942)) and other prop changes can be lost until a new re-render happens.

> Without this change every cell present in view port will be rendered once per scroll stop. If I have large number of rows and columns (100 rows and 10 columns ), and I quickly scroll multiple times (3 times) effectively 100 * 10 * 3 cells will be rendered.

Not sure if I understand this, but If you scroll quickly, I would expect the existing shouldComponenttUpdate to skip render most of times. Or do you mean that you wait before each scroll? If that's the case, we still need to rerender because of the explanation above.

> Once migration to react 16 is done, can use context to provide scroll info to cells.

@wcjordan, we probably might need a new Issue/PR for this.

File: src/FixedDataTableCellGroup.js [hunk 12]
```diff
@@ -194,9 +194,9 @@ class FixedDataTableCellGroup extends React.Component {
 
   shouldComponentUpdate(/*object*/ nextProps) /*boolean*/ {
     return (
-      !nextProps.isScrolling ||
-      this.props.rowIndex !== nextProps.rowIndex ||
-      this.props.left !== nextProps.left
+        (!nextProps.isScrolling && !this.props.isScrolling) ||
```
Comment (pradeepnschrodinger) [inline src/FixedDataTableCellGroup.js]: (same as above comment in FixedDataTableCell.js)

File: src/selectors/columnWidths.js [hunk 13]
```diff
@@ -41,7 +41,7 @@ let columnDefinition;
  * }} The total width of all columns.
  */
 function columnWidths(columnGroupProps, columnProps, scrollEnabledY, width) {
-  const scrollbarSpace = scrollEnabledY ? Scrollbar.SIZE + Scrollbar.OFFSET : 0;
+  const scrollbarSpace = scrollEnabledY ? SCROLLBAR_STYLE.size : 0;
```
Comment (pradeepnschrodinger) [inline src/selectors/columnWidths.js]: Not really sure how the existing logic worked, but if 
`Scrollbar.SIZE + Scrollbar.OFFSET === SCROLLBAR_STYLE.size` in this line,
then wouldn't the lines below where 
`Scrollbar.SIZE` is just replaced with `SCROLLBAR_STYLE.size` without `taking Scrollbar.OFFSET` into account be wrong?

So can you check if the calculations here and below are not wrong (even by 1 pixel, which is the value for Scrollbar.OFFSET)

File: test/selectors/roughHeights-test.js [hunk 14]
```diff
@@ -72,8 +73,8 @@ describe('roughHeights', function() {
 
       assert.deepEqual(result, {
         bufferRowCount: 6,
-        minAvailableHeight: 698,
-        maxAvailableHeight: 698,
+        minAvailableHeight: 800 - 10 - 40 - 35 - 2 - SCROLLBAR_STYLE.size,
```
Comment (pradeepnschrodinger) [inline test/selectors/roughHeights-test.js]: Hey, thanks for describing the tests' result values better! 👍

File: src/FixedDataTableRow.js (hunk 9 above)
Comment (pradeepnschrodinger) [reply src/FixedDataTableRow.js]: > It looks like it might skip those props even if we're not scrolling...

If we take over the approach in [FixedDataTableCellGroup](https://github.com/schrodinger/fixed-data-table-2/blob/v1.0-beta/src/FixedDataTableCellGroup.js#L195), skipping of prop changes will be done while scrolling. This might have a better performance than checking ALL the props, with the cost of loosing updates, but only while scrolling. 

We could also keep the scrolling check, and if it fails we then check for prop changes. This might be a good trade off between both (we'll still loose updates while scrolling though...) This is the approach at FixedDataTableCell.js when pureRendering is turned on.

> We should perhaps instead update this to be a pure component / check across all props.

If this is done, then maybe it can also be done to FixedDataTableCellGroup.js ?

File: src/FixedDataTable.js (hunk 6 above)
Comment (rtrg) [reply src/FixedDataTable.js]: ok, I will get these changes back in that case.

File: src/selectors/columnWidths.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (rtrg) [reply src/selectors/columnWidths.js]: I assumed `offset` is specific to previous scrollbar, let me know if I should restore it back to `size + offset`

File: src/FixedDataTable.js (hunk 10 above)
Comment (rtrg) [reply src/FixedDataTable.js]: For better response to scroll stop, 200ms delay seems too much.

File: src/selectors/columnWidths.js (hunk 13 above)
Comment (rtrg) [reply src/selectors/columnWidths.js]: because I haven't included `offset` even in code, so test cases passing, I will restore that one pixel once you confirm if it is necessary with updated scroll bars

File: src/FixedDataTableRow.js (hunk 9 above)
Comment (rtrg) [reply src/FixedDataTableRow.js]: I will probably make separate PR for `shouldComponentUpdate` changes as suggested by @wcjordan

File: src/FixedDataTable.js (hunk 4 above)
Comment (rtrg) [reply src/FixedDataTable.js]: ok, got your concern here, will make changes!

File: src/selectors/columnWidths.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (wcjordan) [reply src/selectors/columnWidths.js]: I think it is fine to remove the offset as long as we handle the case where the knob is hovered and grows.

**Comment by rtrg** [comment]
> I think the shouldComponentUpdate changes may need some additional discussion. Should we consider splitting those out to a separate PR so we can investigate them further?

Got rid of `shouldComponentUpdate ` changes

> * The build_helpers/cssTransformLoader.js still uses stubs/cssVar.js.  We should restore it before merging.

Done

> * When hovering the scroll knob it grows larger than the available space.  Could you either remove the grow on hover, or update the bar to be large enough to accommodate?

Can you please let me know how to reproduce this? I tried `site-build` and `site-dev-server` to check.

Apologies for the delay, Thanks!

**Comment by wcjordan** [comment]
No worries about the delays, and thanks for the help on this!  Wrt to the hover knob, `yarn run site-dev-server` should turn on an instance at localhost:8080 which you can reproduce on.  On hover the knob is larger than the track which it would be good to fix.  I'm actually also seeing the knob not show up in some cases, so we should investigate that as well.

FixedDataTable

File: src/FixedDataTable.js [hunk 15]
```diff
@@ -1084,12 +1118,12 @@ class FixedDataTable extends React.Component {
       onScrollStart(oldScrollX, oldScrollY, oldFirstRowIndex)
     }
 
-    if (scrollXChanged && onHorizontalScroll) {
-      onHorizontalScroll(scrollX);
+    if (scrollXChanged) {
+      this._scrollToX(scrollX);
```
Comment (pradeepnschrodinger) [inline src/FixedDataTable.js]: This (and below) doesn't seem to work for controlled scrolling... `onVerticalScroll` and `onHorizontalScroll` is not being called.

I think the issue is that setting the internal state with this._scrollbarYRef.scrollTop(y) doesn't trigger a callback. See https://github.com/malte-wessel/react-custom-scrollbars/blob/master/src/Scrollbars/index.js#L97-L211

I'm not even sure if it renders the scroll bar with the updated values? Can you investigate this?

**Comment by pradeepnschrodinger** [comment]
@rtrg , I also see that we can't reach the end of scrolling position in the ColumnGroup and Responsive Resize example (the scrollbar is at the end, but the shadow is still visible). This isn't always reproducible though, but I only see it in this branch, so could be related to https://github.com/schrodinger/fixed-data-table-2/pull/404#discussion_r272800153.

File: src/FixedDataTable.js [hunk 16]
```diff
@@ -1004,10 +1021,29 @@ class FixedDataTable extends React.Component {
...
-        scrollActions.scrollToX(roundedX);
-      }
+      this._scrollToX(roundedX);
+    }
+  }
+
+  /**
+   * This will trigger vertical scroll via scroll bar if present, otherwise triggers via scrollActions directly
+   */
+  _scrollToY(y) {
```
[the remark this answers is not in the document]
Comment (wcjordan) [reply src/FixedDataTable.js]: Ya feel free to file a ticket to clean things up.  We should also consider adding prettier.js and perhaps additional linting rules to ensure any cleanup remains.

**Comment by JoelSanchez** [comment]
Maybe this would help with https://github.com/schrodinger/fixed-data-table-2/issues/456 ?

**Comment by wcjordan** [comment]
@pradeepnschrodinger @rtrg any plans to pick this back up and get it merged?

**Comment by pradeepnschrodinger** [comment]
@wcjordan , this is almost done IIRC.
@rtrg, we can merge this once controlled scrolling is fixed, and the few other minor nits are done.

**Comment by rtrg** [comment]
> @pradeepnschrodinger @rtrg any plans to pick this back up and get it merged?
Hi, few concerns around scroll not reaching to bottom is happening when there are too many number of rows, might be related to react-custom-scrollbars,

Is it ok if I use native scrollbars?

**Comment by pradeepnschrodinger** [comment]
> Hi, few concerns around scroll not reaching to bottom is happening when there are too many number of rows, might be related to react-custom-scrollbars,

Damn... I guess that won't work for our library since we need to handle millions of rows. Any progress on why this occurs or about the fix? 

I just put some debug points and found that the library uses DOM properties directly for calculating offsets and such. 
Specifically, it uses `scrollHeight` on the parent component which renders our child views. 
In this [fiddle](https://jsfiddle.net/NiranjanPBA/w4k1cyzf/5/), it seems that scrollHeight has a max value (`33554431`, which is `2^25 - 1`).
So now the scroll bar reports this clamped value to FDT, and FDT is unable to scroll beyond that....

Hence, it doesn't work for large tables (like our object data example, where the expected value is around `50000000`)...
I'm not sure what the fix is the way we are using it...

> Is it ok if I use native scrollbars?

Sure. Any objections @wcjordan or pointers on what we might face if we go by this route? Some I think are:
* Clicking on the scroll bar to jump instantly won't work out of the box since the native scrollbars seem to slowly go to the clicked position. (FDT goes instantly).
* need 60fps limit (constrained scroll handling)
* don't reach the dead end explained above.

**Comment by wcjordan** [comment]
I'll look a bit deeper at this when I have a chance.  But for inspiration, react-virtualized uses the native scrollbars and handles this using a scaling function which tries to map the scrollbar percentage within those 33.5M pixels onto the percentage scrolled within the full height.  See:
https://github.com/bvaughn/react-virtualized/blob/master/source/Grid/utils/ScalingCellSizeAndPositionManager.js

This is even more important in IE, since there things max out at a much smaller number of pixels.

**Comment by wcjordan** [comment]
Given the new PR for scrollbar extraction, is there anything from this PR which we'd like to keep?  Or can this be closed?

**Comment by rtrg** [comment]
> Given the new PR for scrollbar extraction, is there anything from this PR which we'd like to keep? Or can this be closed?

Closing this, thanks for insightful comments!

I will create new PR on top of custom scrollbars if anything is needed from here.
